### PR TITLE
Set minimum Terraform version to 0.12.26/0.13.0

### PR DESCRIPTION
**Pull request by bmelbourne**
**What this PR does / why we need it:**
As Terraform 0.11 is in the later stages of end-of-life (EOL), the Terraform build/test templates have been updated to be fully compatible with Terraform 0.12.26+ and 0.13.0+.

https://www.hashicorp.com/blog/deprecating-terraform-0-11-support-in-terraform-providers

Minimum Terraform AWS provider is set to v3.12.0 and Google provider is set to v3.44.0.

**Which issue(s) this PR fixes**
N/A

**Special notes for your reviewer:**
The KOPS_FEATURE_FLAGS=Terraform-0.12 will be deprecated in Kops v0.20.

Terraform reference notes regarding 0.12/0.13 provider compatibility.

https://www.terraform.io/docs/configuration/provider-requirements.html#v0-12-compatible-provider-requirements
https://www.terraform.io/upgrade-guides/0-13.html#explicit-provider-source-locations

### CODE REVIEW

**Comment by hakman** [comment]
Something's not quite right with the commits.

**Comment by bmelbourne** [comment]
> Something's not quite right with the commits.

It's WIP...I just needed to check the Virgin Media Web Safe settings weren't going to mess with the commits in my new branch between my forked repo and the kops upstream repo.

Just running the various tests locally and then I'll squash the commits and final rebase! 😉

**Comment by hakman** [comment]
Ah, cool. :)
/ok-to-test

File: docs/terraform.md [hunk 1]
```diff
@@ -54,7 +49,32 @@ $ kops create cluster \
   --target=terraform
 ```
 
-The above command will create kops state on S3 (defined in `--state`) and output a representation of your configuration into Terraform files. Thereafter you can preview your changes and then apply as shown below:
+The above command will create kops state on S3 (defined in `--state`) and output a representation of your configuration into Terraform files. Thereafter you can preview your changes in `kubernetes.tf` and then use Terraform to create all the resources as shown below:
+
+Additional Terraform `.tf` files could be added at this stage to custom your deployment, but remember the kops state should continue to remain the ultimate source of truth for the Kubernetes cluster.
```
Comment (rifelpet) [inline docs/terraform.md:54]:
```suggestion
Additional Terraform `.tf` files could be added at this stage to customize your deployment, but remember the kops state should continue to remain the ultimate source of truth for the Kubernetes cluster.
```

File: upup/pkg/fi/cloudup/terraform/target_hcl2.go [hunk 2]
```diff
@@ -87,7 +84,22 @@ func (t *TerraformTarget) finish012(taskMap map[string]fi.Task) error {
...
+
+	if t.Cloud.ProviderID() == kops.CloudProviderGCE {
+		writeMap(requiredProvidersBody, "google", map[string]cty.Value{
+			"source":  cty.StringVal("hashicorp/google"),
+			"version": cty.StringVal(">= 3.44.0"),
+		})
+	} else if t.Cloud.ProviderID() == kops.CloudProviderAWS {
+		writeMap(requiredProvidersBody, "aws", map[string]cty.Value{
+			"source":  cty.StringVal("hashicorp/aws"),
+			"version": cty.StringVal(">= 3.12.0"),
```
Comment (rifelpet) [inline upup/pkg/fi/cloudup/terraform/target_hcl2.go:100]: By increasing the version constraints on the aws and google providers we're forcing anyone to upgrade their provider version when they upgrade to this version of kops. Is that really necessary? Ideally we could find the oldest 3.X provider version that supports all of the functionality that kops uses and use that in the version constraint.

I know I've had issues upgrading aws provider minor versions in the past but not with any resource types used by kops.

Just from memory, I think one of the most recent changes to kops' terraform output that requires a recent provider version would be https://github.com/kubernetes/kops/pull/9533 and [2.36.0](https://github.com/terraform-providers/terraform-provider-aws/blob/release/2.x/CHANGELOG.md#2460-january-23-2020) but I don't recall if newer terraform versions require aws provider 3.X.

**Review by rifelpet — COMMENTED** [review]
lgtm overall, just a couple comments.
https://github.com/kubernetes/kops/pull/10101 creates the 1.20 release notes and mentions the feature flag removal but if we do end up requiring aws provider 3.X then we should probably mention that and link to the [upgrade guide](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade) as well since it has additional instructions.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade#provider-version-configuration

File: upup/pkg/fi/cloudup/terraform/target_hcl2.go (hunk 2 above)
Comment (bmelbourne) [reply upup/pkg/fi/cloudup/terraform/target_hcl2.go:100]: Agreed. Probably best not to force too much change on developers using Kops v1.20.

After some testing the minimum provider versions to create a valid Terraform plan consisting of a simple Kops cluster is `2.46.0` for AWS and `2.19.0` for Google.

AWS Terraform provider `2.45.0` generates the following error...

```
Error: Unsupported argument

  on kubernetes.tf line 292, in resource "aws_key_pair" "kubernetes-training-kops-k8s-local-3e19983429436e7b1ff18156a91241c6":
 292:   tags = {

An argument named "tags" is not expected here.
```

Relevant scripts updated and ready for review.

**Comment by rifelpet** [comment]
/lgtm
/approve

Thanks for doing this! Feel free to `/hold cancel` and remove WIP from the PR title to get this merged.

**Comment by bmelbourne** [comment]
> /lgtm
> /approve
> 
> Thanks for doing this! Feel free to `/hold cancel` and remove WIP from the PR title to get this merged.

@rifelpet thanks! 😄 
I was just waiting for the CI tests to complete before removing the labels.

**Pull request merged**